src/Modules/GravityRing.cs

```diff
@@ -36,30 +37,73 @@ public override void OnStart(StartState state)
 
 			// get animations
 			deploy_anim = new Animator(part, deploy);
-			rotate_anim = new Animator(part, rotate);
 
-			// if is using Transform
-			rotate_transf = new Transformator(part, rotate, SpinRate, SpinAccelerationRate);
+			if (rotateIsTransform) rotate_transf = new Transformator(part, rotate, SpinRate, SpinAccelerationRate);
+			else rotate_anim = new Animator(part, rotate);
 
 			// set animation state / invert animation
 			deploy_anim.still(deployed ? 1.0f : 0.0f);
 			deploy_anim.stop();
 
-			if (deployed)
+			Update();
+		}
+
+		public bool is_rotating()
+		{
+			if (rotateIsTransform)
+			{
+				return rotate_transf.IsRotating() && !rotate_transf.IsStopping();
+			}
+			else
+			{
+				return rotate_anim.playing();
+			}
+		}
+
+		private void set_rotation(bool rotate)
+		{
+			if (rotate)
+			{
+				if (rotateIsTransform) rotate_transf.Play();
+				else
+				{
+					rotate_anim.resume(false);
+					if (!rotate_anim.playing()) rotate_anim.play(false, true);
+				}
+			}
+			else
 			{
-				rotate_transf.Play();
-				rotate_anim.play(false, true);
+				if (rotateIsTransform) rotate_transf.Stop();
+				else rotate_anim.pause();
 			}
+		}
 
-			// show the deploy toggle if it is deployable
-			Events["Toggle"].active = deploy.Length > 0;
+		bool should_start_rotation()
+		{
+			return (isHabitat && deployed) || (!isHabitat && !deploy_anim.playing());
+		}
+
+		bool is_consuming_energy()
+		{
+			if (deploy_anim.playing())
+			{
+				return true;
+			}
+			if (rotateIsTransform)
+			{
+				return rotate_transf.IsRotating() && !rotate_transf.IsStopping();
+			}
+			else
+			{
+				return rotate_anim.playing();
+			}
 		}
 
 		public void Update()
 		{
 			// update RMB ui
 			Events["Toggle"].guiName = deployed ? Localizer.Format("#KERBALISM_Generic_RETRACT") : Localizer.Format("#KERBALISM_Generic_DEPLOY");
-			Events["Toggle"].active = deploy.Length > 0 && !deploy_anim.playing() && !waitRotation && ResourceCache.Info(vessel, "ElectricCharge").amount > ec_rate;
+			Events["Toggle"].active = (deploy.Length > 0) && (part.FindModuleImplementing<Habitat>() == null) && !deploy_anim.playing() && !waitRotation && ResourceCache.Info(vessel, "ElectricCharge").amount > ec_rate;
 		}
 
 		public void FixedUpdate()
```

Comment:
Might be better for performance issues to move code in the `FixedUpdate()` method into the `Update()` method.
`FixedUpdate()` should only be used for physics calculation math or time sensitive data that ***MUST*** be calculated every physics frame. Code in `FixedUpdate()` should be kept as short and optimized as possible. (see the `Profiler.cs` file and note the code that is called in `FixedUpdate()` is the code used for detecting the length of execution times only). 

A lot of people misunderstand the differences of the uses of these two methods and dump everything in `FixedUpdate()` because it is called every physics frame. 😉

Reply:
One day soon I will be going through all the files in Kerbalism and moving code that is not physics frame sensitive from `FixedUpdate()` to `Update()`